src/protocol/connection.rs

```diff
@@ -0,0 +1,86 @@
+use futures::TryFutureExt;
+use obfswire::{Config, ObfuscatedStream, SharedKey};
+use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+use tokio_rustls::{client, rustls::pki_types::ServerName, server, TlsAcceptor, TlsConnector};
+use tokio_util::compat::{TokioAsyncReadCompatExt, TokioAsyncWriteCompatExt};
+
+use crate::{
+    packet_stream::{PacketReceiver, PacketSender, TaggedPacketReceiver, TaggedPacketSender},
+    protocol::NetworkConfig,
+};
+
+pub struct Connection<Stream: Send>(Stream);
+
+impl<Stream> Connection<Stream>
+where
+    Stream: AsyncRead + AsyncWrite + Unpin + Send,
+{
+    pub fn new(stream: Stream) -> Self {
+        Self(stream)
+    }
+
+    pub async fn connect_tls(
+        self,
+        connector: &TlsConnector,
+        domain: ServerName<'static>,
+    ) -> std::io::Result<Connection<client::TlsStream<Stream>>> {
+        connector
+            .connect(domain, self.0)
+            .map_ok(Connection::new)
+            .await
+    }
+
+    pub async fn accept_tls(
+        self,
+        acceptor: &TlsAcceptor,
+    ) -> std::io::Result<Connection<server::TlsStream<Stream>>> {
+        acceptor.accept(self.0).map_ok(Connection::new).await
+    }
+
+    pub async fn start_obfs_server(
+        mut self,
+    ) -> std::io::Result<Connection<ObfuscatedStream<Stream>>> {
+        let key = SharedKey::from_entropy();
+        self.0.write_all(key.as_bytes()).await?;
+        let config = Config::builder_with_shared_key(key)
+            .with_default_cipher()
+            .no_padding();
+        Ok(Connection::new(ObfuscatedStream::with_config_in(
+            config, self.0,
+        )))
+    }
+
+    pub async fn start_obfs_client(
+        mut self,
+    ) -> std::io::Result<Connection<ObfuscatedStream<Stream>>> {
+        // TODO: share key in a safe manner
```

Comment:
blocking on this in [another pr](https://github.com/Mad-Proger/opaque-vpn/pull/8)